### PR TITLE
nexttrace: init at 1.1.3

### PR DESCRIPTION
**Pull request by SharzyL**
###### Description of changes

Nexttrace (https://github.com/sjlleo/nexttrace) is an open source visual route tracking CLI tool. 

It requires a nixos module to add setcap wrapper.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/programs/nexttrace.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  options = {
+    programs.nexttrace = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = lib.mdDoc ''
+          Whether to add nexttrace to the global environment and configure a
+          setcap wrapper for it.
+        '';
+      };
```
Comment (Kranzes) [inline nixos/modules/programs/nexttrace.nix:19]:
```suggestion
      enable = mkEnableOption (lib.mdDoc "Nexttrace to the global environment and configure a setcap wrapper for it");
```

File: nixos/modules/programs/nexttrace.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.nexttrace;
+        defaultText = literalExpression "pkgs.nexttrace";
+        description = lib.mdDoc ''
+          The package to use.
+        '';
+      };
```
Comment (Kranzes) [inline nixos/modules/programs/nexttrace.nix:28]:
```suggestion
      package = mkPackageOptionMD pkgs "nexttrace" { };
```

**Review by Kranzes — CHANGES_REQUESTED** [review]

File: pkgs/tools/networking/nexttrace/default.nix [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  };
+  vendorSha256 = "sha256-SA22qjPS9c5MDKBA0YpL6BmFp/7/RGX0NYjNiKce6rM=";
+  doCheck = false; # its tests require networking
+
+  meta = with lib; {
+    description = "An open source visual route tracking CLI tool";
+    homepage = "https://mtr.moe";
+    license = licenses.gpl3;
+    maintainers = with maintainers; [ sharzy ];
+    platforms = platforms.unix;
```
Comment (Kranzes) [inline pkgs/tools/networking/nexttrace/default.nix:20]:
```suggestion
    platforms = platforms.linux;
```
If this requires a module shouldn't this be set to `linux`?

**Comment by Kranzes** [comment]
I recommend using this expression (mostly generated by [nix-init](https://github.com/nix-community/nix-init)):
```nix
{ lib
, buildGoModule
, fetchFromGitHub
}:

buildGoModule rec {
  pname = "nexttrace";
  version = "1.1.3";

  src = fetchFromGitHub {
    owner = "sjlleo";
    repo = pname;
    rev = "v${version}";
    hash = "sha256-sOTQBh6j8od24s36J0e2aKW1mWmAD/ThfY6pd1SsSlY=";
  };

  vendorHash = "sha256-ckGoDV4GNp0mG+bkCKoLBO+ap53R5zrq/ZSKiFmVf9U=";

  doCheck = false; # Tests require a network connection.

  ldflags = [
    "-s"
    "-w"
    "-X github.com/xgadget-lab/nexttrace/printer.version=v${version}"
  ];

  meta = with lib; {
    description = "An open source visual route tracking CLI tool";
    homepage = "https://mtr.moe";
    license = licenses.gpl3Only;
    maintainers = with maintainers; [ sharzy ];
    inherit (platforms) linux;
  };
}
```
In addition you should change the commit and PR name to `nexttrace: init at 1.1.3`, and for the nixos module commit just rename it to `nixos/nexttrace: init`.

**Comment by Kranzes** [comment]
For the module we can also be a bit more explicit without much added effort since this is a small module:
```nix
{ config, lib, pkgs, ... }:

let
  cfg = config.programs.nexttrace;
in
{
  options.programs.nexttrace = {
    enable = lib.mkEnableOption (lib.mdDoc "Nexttrace to the global environment and configure a setcap wrapper for it");
    package = lib.mkPackageOptionMD pkgs "nexttrace" { };
  };

  config = lib.mkIf cfg.enable {
    environment.systemPackages = [ cfg.package ];

    security.wrappers.nexttrace = {
      owner = "root";
      group = "root";
      capabilities = "cap_net_raw,cap_net_admin+eip";
      source = "${lib.getExe cfg.package}";
    };
  };
}
```

File: pkgs/tools/networking/nexttrace/default.nix (hunk 3 above)
Comment (SharzyL) [reply pkgs/tools/networking/nexttrace/default.nix:20]: For non-nixos user, the user can call nexttrace with `sudo` or something. Similar tools, e.g. `mtr`, `inettools`, also take `platforms.unix`. Hence I think `unix` is ok.

**Comment by SharzyL** [comment]
Now it is upgraded and made explicit.

**Comment by Kranzes** [comment]
> Now it is upgraded and made explicit.

Please use the expressions I sent for the module and the package, they have more improvements. Also adjust the commit of the nixos module to `nixos/nexttrace: init`.

**Comment by SharzyL** [comment]
```nix
{
  meta = with lib; {
    description = "An open source visual route tracking CLI tool";
    homepage = "https://mtr.moe/";
    license = licenses.gpl3Only;
    maintainers = with maintainers; [ sharzy ];
  };
}
```

is the missing `platforms =` intended, or should I add `platforms = platforms.unix`?

**Comment by Kranzes** [comment]
It's intended since it's the default.

**Comment by SharzyL** [comment]
updated

**Pull request merged**